src/core/entity.js

```diff
@@ -124,13 +132,16 @@ class Entity {
   }
 
   rotate_along(vec, rad) {
-    // Negate the origin vector so that we can go back.
-    const reverse_origin = this.origin.map(i => -i);
-    const rotation_matrix = math.mat4.create();
-    math.mat4.from_rotation(rotation_matrix, rad, vec);
-    math.mat4.translate(this.local_matrix, this.local_matrix, this.origin);
-    math.mat4.multiply(this.local_matrix, this.local_matrix, rotation_matrix);
-    math.mat4.translate(this.local_matrix, this.local_matrix, reverse_origin);
+    const rotation = math.quat.create();
+    math.quat.set_axis_angle(rotation, vec, rad);
+
+    // Quaternion multiplication: A * B applies the A rotation first, B second,
+    // relative to the coordinate system resulting from A.
+    math.quat.multiply(rotation, this.rotation, rotation);
+    this._rotation = rotation;
+    math.mat4.compose(
+      this.matrix, rotation, this.position, this.scale, this.origin
```

Comment:
Something's wrong for non-zero origins:
![pos-1](https://user-images.githubusercontent.com/347657/29752650-c3fb152c-8b62-11e7-9c4e-bff10a0c8228.gif)